zipline/pipeline/loaders/events.py

```diff
@@ -0,0 +1,126 @@
+from abc import abstractmethod
+
+import numpy as np
+import pandas as pd
+from six import iteritems
+from toolz import merge
+
+from .base import PipelineLoader
+from .frame import DataFrameLoader
+from .utils import next_date_frame, previous_date_frame, previous_value
+
+TS_FIELD_NAME = "timestamp"
+
+
+class EventsLoader(PipelineLoader):
+    """
+    Abstract loader.
+
+    Does not currently support adjustments to the dates of known events.
+
+    Parameters
+    ----------
+    all_dates : pd.DatetimeIndex
+        Index of dates for which we can serve queries.
+    events_by_sid : dict[int -> pd.DataFrame]
+        Dict mapping sids to DataFrames representing dates on which events
+        occurred along with other associated values.
+
+        If the DataFrames contain a "timestamp" column, that column is
+        interpreted as the date on which we learned about the event.
+
+        If the DataFrames do not contain a "timestamp" column, we assume we
+        knew about the event on all prior dates.  This mode is only supported
+        if ``infer_timestamp`` is explicitly passed as a truthy value.
+
+    infer_timestamps : bool, optional
+        Whether to allow omitting the "timestamp" column.
+    """
+
+    def __init__(self,
+                 all_dates,
+                 events_by_sid,
```

Comment:
noting what we talked about irl: add a frozenset which says what the expected columns are. The constructor will validate the shape of the dataframes passed to fail early with a good error message. If the order of this set is one, then we can coerce input from a datetimeindex or series.